bst.py

```diff
@@ -83,13 +107,76 @@ def balance(self):
         negative balance if the right side is deeper, and zero if
         the tree is balanced
         """
-        if self.left is None and self.right is None:
-            return 0
-        if self.left is None:
-            return -1 * self.right.depth()
-        if self.right is None:
-            return self.left.depth()
-        return self.left.depth() - self.right.depth()
+        l = self.left.depth() if self.left else 0
+        r = self.right.depth() if self.right else 0
+        return l - r
+
+    # Depth first traverals
+    def pre_order(self):
+        """
+        Call a pre-order depth first traversal generator.
+        Return nodes in turn from generator.
+        """
+        parent_stack = []
+        while parent_stack or self is not None:
+            if self is not None:
```

Comment:
This doesn't need to be here. If self is None... It wouldn't even be in this while loop...